contrib/IECoreUSD/src/IECoreUSD/VolumeAlgo.cpp

```diff
@@ -0,0 +1,201 @@
+
+#include "IECoreUSD/ObjectAlgo.h"
+
+#include "IECoreVDB/VDBObject.h"
+
+#include "IECore/MessageHandler.h"
+
+IECORE_PUSH_DEFAULT_VISIBILITY
+#include "pxr/usd/usdVol/volume.h"
+#include "pxr/usd/usdVol/openVDBAsset.h"
+IECORE_POP_DEFAULT_VISIBILITY
+
+using namespace pxr;
+using namespace IECore;
+using namespace IECoreScene;
+using namespace IECoreUSD;
+
+// Overview
+// ========
+//
+// The closest analogue to `UsdVolVolume` in Cortex is `IECoreVDB::VDBObject`,
+// but the two classes have significant differences.
+//
+// - UsdVol provides schemas for referencing volume files on disk, but currently
+//   provides no access to the volume data itself. On the other hand, VDBObject
+//   provides direct access to data via `openvdb::GridBase::Ptr`, making it
+//   suitable for use in live volume generation and processing. We take
+//   advantage of this in Gaffer by providing various nodes for manipulating
+//   volumes.
+// - UsdVolVolume allows the referencing of multiple fields (grids in VDB
+//   parlance) from multiple different files, and allows the fields to be given
+//   names that are distinct from the grid names themselves. While a VDBObject
+//   _can_ be composed of multiple grids from arbitrary sources, this is done by
+//   calling `insertGrid()`, and doesn't track the file of origin (if there even
+//   was one). VDBObject is only guaranteed to reference file-backed data when
+//   constructed from a single filename and when the loaded grids have not been
+//   modified subsequently.
+//
+// While there is scope for extending VDBObject to provide a cleaner mapping, at
+// present this would have limited benefit. All our supported VDB-consuming
+// renderers have volume objects that reference only a single file, so we would
+// just be pushing the mismatch further down the pipeline. So for now we provide
+// the closest mapping we can, and issue warnings for any loss of data.
+
+// Reading
+// =======
+
+namespace
+{
+
+IECore::ObjectPtr readVolume( pxr::UsdVolVolume &volume, pxr::UsdTimeCode time, const Canceller *canceller )
+{
+	std::string fileName;
+	for( const auto &[fieldName, fieldPath] : volume.GetFieldPaths() )
+	{
+		UsdVolOpenVDBAsset fieldAsset( volume.GetPrim().GetPrimAtPath( fieldPath ) );
+		if( !fieldAsset )
+		{
+			IECore::msg(
+				IECore::Msg::Warning, "IECoreVDB::VolumeAlgo::readVolume",
+				"Ignoring \"" + fieldPath.GetAsString() + "\" because it is not an OpenVDBAsset"
+			);
+			continue;
+		}
+
+		SdfAssetPath fieldAssetPath;
+		fieldAsset.GetFilePathAttr().Get( &fieldAssetPath, time );
+		const std::string fieldFileName = fieldAssetPath.GetResolvedPath();
+
+		if( fileName.empty() )
+		{
+			fileName = fieldFileName;
+		}
+		else if( fieldFileName != fileName )
+		{
+			IECore::msg(
+				IECore::Msg::Warning, "IECoreVDB::VolumeAlgo::readVolume",
+				"Ignoring file \"" + fieldFileName + "\" from field \"" + fieldPath.GetAsString() + "\""
+			);
+		}
+	}
```

Comment:
If I'm understanding this correctly, one could have two USD files, one of which reads one "field" ( grid ) from a VDB with 4 grids, and one of which reads all 4 grids by listing 4 fields. Currently, we would load both of these USD files exactly the same, without warning in either case? That feels like a slight contradiction to your stated principle of "we warn for any loss of data", though I'm not sure what the correct warning would be.